[MRG+1] Adding multi output checks to common tests

rok opened the pull request:
#### Reference Issues/PRs

Fixes #13187.

#### Changes

This implements new classifier and regressor checks to test for multi-output support in common tests.
Some random forest and tree classifier test are removed as they are duplicating this functionality.

rok committed:
Adding multi-output checks for estimators.

[MRG] Adding multi output checks to common tests (#13187)

Removing redundant tests.
Adding tests to check_classifier_multioutput and check_regressor_multioutput.

TomDLT commented on sklearn/utils/estimator_checks.py:
```diff
@@ -57,12 +57,36 @@
 BOSTON = None
 CROSS_DECOMPOSITION = ['PLSCanonical', 'PLSRegression', 'CCA', 'PLSSVD']
 
+NOT_MULTI_OUTPUT = ["ARDRegression", "AdaBoostRegressor", "BayesianRidge",
```
We probably don't need that.

rok replied:
Ok.

TomDLT commented on sklearn/utils/estimator_checks.py:
```diff
@@ -57,12 +57,36 @@
+                    "RidgeClassifier", "RidgeClassifierCV", "SGDClassifier",
+                    "SVC"]
+
+ALLOW_NAN = ['Imputer', 'SimpleImputer', 'MissingIndicator',
```
We probably don't need that.

TomDLT commented on sklearn/utils/estimator_checks.py:
```diff
@@ -123,6 +147,9 @@ def _yield_classifier_checks(name, classifier):
     yield check_classifiers_one_label
     yield check_classifiers_classes
     yield check_estimators_partial_fit_n_features
+    # test classifiers for multi-target output
```
I don't find this comment more informative that the name of the function `check_classifier_multioutput`.
You might remove it.

TomDLT commented on sklearn/utils/estimator_checks.py:
```diff
@@ -171,6 +198,9 @@ def _yield_regressor_checks(name, regressor):
     yield partial(check_regressors_train, readonly_memmap=True)
     yield check_regressor_data_not_an_array
     yield check_estimators_partial_fit_n_features
+    # test regressors for multi-target output
```
Same here.

TomDLT commented on sklearn/utils/estimator_checks.py:
```diff
@@ -123,6 +147,9 @@ def _yield_classifier_checks(name, classifier):
     yield check_classifiers_classes
     yield check_estimators_partial_fit_n_features
+    # test classifiers for multi-target output
+    if name not in NOT_MULTI_OUTPUT:
```
Now that we have merged #8022, you should use tags instead.
(`if tags["multi_output"]: ...`)

You should also update `doc/developers/contributing.rst` which says `multioutput - unused for now`.

rok replied:
Done. Tags are great!

rok replied:
I had to add some exceptions to make this work, see [commit](https://github.com/scikit-learn/scikit-learn/pull/13392/commits/b720e066858c735223ed8a21e3631ac6233b3b55).

It appears that in case of [Lars](https://github.com/scikit-learn/scikit-learn/blob/fec7670ad07d4e639995ad005796021928344805/sklearn/linear_model/least_angle.py#L760) the parent class and some subclasses (Lars, LassoLars) support multioutput while some of it's subclasses (LarsCV, LassoLarsCV, LassoLarsIC) do not.

I removed `MultiOutputMixin` from Lars class and excluded Lars and LassoLars from [`check_supervised_y_2d`](https://github.com/scikit-learn/scikit-learn/blob/b73a51bcda362d94d8907915a382a8eb403554c8/sklearn/utils/estimator_checks.py#L1688) check as Lars and LassoLars were not throwing `DataConversionWarning` which causes `check_supervised_y_2d` to fail.

rok replied:
@TomDLT do you think this would be ok?

TomDLT commented on sklearn/neighbors/regression.py:
```diff
@@ -17,12 +17,13 @@
 
 class KNeighborsRegressor(NeighborsBase, KNeighborsMixin,
                           SupervisedFloatMixin,
+                          RequiresPositiveDataMixin,
```
Why did you add this here ? `KNeighborsRegressor` does not require positive data in the default case.

rok replied:
Yes, that was not the best idea. I did it to fix [this test](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/tests/test_estimator_checks.py#L481). I see now it is using `KNeighborsRegressor(metric='precomputed')` and failing, while `KNeighborsRegressor()` passes. I'll remove the metric parameter here.

TomDLT commented on sklearn/utils/estimator_checks.py:
```diff
@@ -170,8 +173,10 @@ def _yield_regressor_checks(name, regressor):
+        yield check_regressor_multioutput
     yield check_regressors_no_decision_function
-    if not tags["no_validation"]:
+    if not tags["no_validation"] and name not in ["Lars", "LassoLars"]:
```
I don't think this is ideal, we should fix the tag in `Lars` instead.

Also, why is it linked to `tags["no_validation"]` ?

rok replied:
After fixing mixins this is no longer needed to pass tests. Removing.

TomDLT commented on sklearn/base.py:
```diff
@@ -555,6 +555,12 @@ def _more_tags(self):
 
 
+class RequiresPositiveDataMixin(object):
+    """Mixin to mark estimators that requres positive data."""
```
requres -> require

rok replied:
Done.

TomDLT commented on sklearn/linear_model/coordinate_descent.py:
```diff
@@ -1040,7 +1040,8 @@ def _path_residuals(X, y, train, test, path, path_params, alphas=None,
 
 
-class LinearModelCV(LinearModel, MultiOutputMixin, metaclass=ABCMeta):
+class LinearModelCV(LinearModel,
```
You should keep it in one line.

TomDLT commented on sklearn/linear_model/least_angle.py:
```diff
@@ -757,7 +757,7 @@ def _lars_path_solver(X, y, Xy=None, Gram=None, n_samples=None, max_iter=500,
 # Estimator classes
 
-class Lars(LinearModel, RegressorMixin, MultiOutputMixin):
+class Lars(LinearModel, RegressorMixin):
```
Is there a way to keep the mixin here, but to override the tag in subclasses that do not handle multioutput ?

rok replied:
I now solved it with intermediate classes so we now have `_baseLars` without `MultiOutputMixin` and `Lars` class that inherits `_baseLars` and has `MultiOutputMixin`.

The way tags are implemented does not allow for mixins that negate other mixins. Probably it wouldn't be a good idea anyway.

TomDLT reviewed (commented):
Not sure about these tag changes (https://github.com/scikit-learn/scikit-learn/pull/13392#discussion_r263429684). ping @amueller

TomDLT commented on sklearn/utils/tests/test_estimator_checks.py:
```diff
@@ -471,14 +471,14 @@ def __init__(self, you_should_set_this_=None):
 
     # test precomputed metric
-    est = KNeighborsRegressor(metric='precomputed')
+    est = KNeighborsRegressor()
```
You should not remove `metric='precomputed'`, which is essential to this test.
Why is `test_check_estimator_pairwise` related to your new multioutput tests?

rok replied:
Hey @TomDLT, thanks for the super fast response! :)

If I set `metric='precomputed'` I get `ValueError: Negative values in data passed`.

Originally I fixed it by adding `RequiresPositiveDataMixin` to the `KNeighborsRegressor` and then by removing `metric='precomputed'`. After reading deeper I agree neither is a good idea.

How about only setting the `requires_positive_data` tag if `self.metric == 'precomputed'`? It feels kinda hacky but solves it. See latest commit.

TomDLT replied:
Oh I see, it comes from the `predict` in `check_regressor_multioutput`, which checks that precomputed distances are non-negative, whereas the `fit` does not check (but probably should).

I don't see any obvious reason to protest against your conditional tag:
```py
    def _more_tags(self):
        if self.metric == 'precomputed':
            return {'requires_positive_data': True}
        else:
            return {}
```
but I would prefer to have some input from other people involved in estimator tags.
@amueller @jnothman @rth @glemaitre.

rok replied:
Any further comments @amueller @jnothman @rth @glemaitre @TomDLT or can we merge?

TomDLT reviewed (changes requested):
Thanks for the update !
I like much more your Mixin fix now.

You will also need to add en entry in https://github.com/scikit-learn/scikit-learn/blob/master/doc/whats_new/v0.21.rst#changes-to-estimator-checks, writing the change, this PR number and your GitHub name.

jnothman commented:
I wonder if requires_positive_data is implied by accepting pairwise data in
all cases.

jnothman commented:
But there's certainly nothing wrong with conditional tags

rok commented:
> I wonder if requires_positive_data is implied by accepting pairwise data in all cases.

@jnothman - right, is there some other way to know?

TomDLT commented on doc/whats_new/v0.21.rst:
```diff
@@ -523,3 +523,14 @@ These changes mostly affect library developers.
+  :func:`~utils.estimator_checks._yield_regressor_checks` respectively.
+  Added three intermediate private classes (
+  :class:`~linear_model.coordinate_descent._BaseLinearModelCV`,
+  :class:`~linear_model.least_angle._BaseLars`,
```
This whatsnew entry is rather verbose.
I would remove the mention to `_yield_classifier_checks` and the mention to private classes. These are implementation details that should not affect users.

rok replied:
Changed! :)

TomDLT reviewed (approved):
LGTM

We just need to wait for a second review.

thomasjpfan commented on sklearn/ensemble/tests/test_forest.py:
```diff
@@ -1337,25 +1337,3 @@ def test_backend_respected():
-    y = np.vstack([
-        iris.target.astype(float),
-        iris.target.astype(int),
-        iris.target.astype(str),
```
The original test, uses y with dtype `<U32` which contains strings. The new common test only accounts for numeric types.

rok replied:
Good point. I added that check again for estimators that don't use LabelBinarizer.

TomDLT commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1296,6 +1301,33 @@ def check_estimators_partial_fit_n_features(name, estimator_orig):
+
+    estimator_orig = clone(estimator_orig)
+    X, y = make_multilabel_classification(random_state=42)
+    if not (hasattr(estimator_orig, '_label_binarizer') or not
```
Why do you need this condition? I would rather name explicitly the estimators that do not handle multioutput string labels (`if name in ["Foo", "Bar]:`). Then we might open an issue to solve these exceptions.

rok replied:
Good point! I've changed the condition and pushed. :)
The issue appears to be how `LabelBinarizer` is used by `OneVsRestClassifier` and `RidgeClassifierCV`.
In both cases this error is raised:
```
>           raise ValueError("Multioutput target data is not supported with "
                             "label binarization")
E           ValueError: Multioutput target data is not supported with label binarization
```

rok commented:
Ping! :)

rok commented:
I've rebased and maybe we could get it merged in v0.22? :)

rok commented:
Rebased and fixed new issues. Since things changed it might be good to do another review.

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1323,6 +1327,32 @@ def check_estimators_partial_fit_n_features(name, estimator_orig):
+    X, y = make_multilabel_classification(random_state=42)
+    if name not in ['OneVsRestClassifier', 'RidgeClassifierCV']:
+        estimator_orig.fit(X, y.astype(str)).predict(X)
+    estimator_orig.fit(X, y).predict(X)
```
Can we check the shape at least to some degree?
Also, I'd really like to avoid hard-coding any estimators in the tests.

Why are we not testing ``decision_function`` and ``predict_proba`` here?

rok replied:
I added some checks. Please review.

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1323,6 +1327,32 @@ def check_estimators_partial_fit_n_features(name, estimator_orig):
+    if tags["requires_positive_X"]:
+        X = np.abs(X)
+
+    if name == "RANSACRegressor":
```
I would really try to avoid this if possible at all.

rok replied:
Removed.

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1323,6 +1327,32 @@ def check_estimators_partial_fit_n_features(name, estimator_orig):
+    if name == "RANSACRegressor":
+        estimator_orig.min_samples = 99
+
+    estimator_orig.fit(X, y).predict(X)
```
We should check the shape of the output.

rok replied:
Added checks, please review.

amueller commented on sklearn/utils/tests/test_estimator_checks.py:
```diff
@@ -422,7 +422,10 @@ def test_check_estimator():
 
     # Check regressor with requires_positive_y estimator tag
-    check_estimator(RequiresPositiveYRegressor)
+    ValueError('negative y values not supported!')
```
This line is not doing anything, right?

rok replied:
Removed.

amueller reviewed (commented):
looks reasonable overall, but I'd add more tests. Right now it's only smoke tests.
Also I'd really like to avoid any hardcoding within the tests.

jnothman commented on doc/whats_new/v0.22.rst:
```diff
@@ -299,3 +299,15 @@ These changes mostly affect library developers.
+- Created two new multioutput estimator tests:
+  :func:`~utils.estimator_checks.check_classifier_multioutput` and
+  :func:`~utils.estimator_checks.check_regressor_multioutput`.
+  Added them to :func:`~utils.estimator_checks._yield_classifier_checks` and
```
This is not public API. Basically they are included in check_estimator

rok replied:
Got it. Shortened it, please review.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1346,6 +1350,85 @@ def check_estimators_partial_fit_n_features(name, estimator_orig):
+
+    assert y_pred.dtype == np.dtype('float')
+    assert y_pred.shape == y.shape
+    assert score.dtype == np.dtype('float')
```
This is not testing the regressor, but the metric

rok replied:
Indeed, removing.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1346,6 +1350,85 @@ def check_estimators_partial_fit_n_features(name, estimator_orig):
+    assert y_pred.dtype == np.dtype('float')
+    assert y_pred.shape == y.shape
+    assert score.dtype == np.dtype('float')
+    assert score.shape == (n_targets,)
```
Ditto

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1346,6 +1350,85 @@ def check_estimators_partial_fit_n_features(name, estimator_orig):
+    n_features = n_samples
+    n_targets = 5
+
+    if _is_pairwise_metric(estimator):
```
Why not use pairwise_estimator_convert_X? Some estimators may require pairwise X to be a valid distance or kennel matrix, which you cannot ensure using your current approach

rok replied:
Good point. Replaced.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1346,6 +1350,85 @@ def check_estimators_partial_fit_n_features(name, estimator_orig):
+            assert_array_equal(rankdata(a), rankdata(b[:, i]))
+
+    scores = jaccard_score(y_pred, y, average=None)
+    assert scores.dtype == np.dtype('float')
```
This is not testing the estimator, but the metric

rok replied:
Removed.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1346,6 +1350,85 @@ def check_estimators_partial_fit_n_features(name, estimator_orig):
+
+    if (hasattr(estimator, "decision_function") and
+            hasattr(estimator, "predict_proba")):
+        for i in range(n_labels):
```
Hmm .. we currently have two different ways of representing probabilities for multilabel classifiers. I think this needs some discussion

rok replied:
Agreed. I propose we open a separate issue for that.

rok replied:
@jnothman thanks for the review! :)

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1346,6 +1350,85 @@ def check_estimators_partial_fit_n_features(name, estimator_orig):
+            b = estimator.decision_function(X)
+            assert_array_equal(rankdata(a), rankdata(b[:, i]))
+
+    scores = jaccard_score(y_pred, y, average=None)
```
There's no point calculating scores and not using them

rok replied:
Removing.

jnothman reviewed (commented):
Sorry if I've missed some pertinent discussion. @amueller, is it time to come to terms with the different multilabel probability formats in OneVsRestClassifier (2d) and generic multi output models (3d)?

glemaitre reviewed (approved):
The tests look good to me. Some nitpick to be addressed.
However, I don't have enough background and running miles to comment on the 2D/3D output format https://github.com/scikit-learn/scikit-learn/pull/13392#pullrequestreview-269920559

@amueller It would be nice to have your thoughts on this.

amueller commented on sklearn/neighbors/regression.py:
```diff
@@ -146,6 +146,12 @@ def __init__(self, n_neighbors=5, weights='uniform',
               metric_params=metric_params, n_jobs=n_jobs, **kwargs)
         self.weights = _check_weights(weights)
 
+    def _more_tags(self):
```
Shouldn't this be handled by ``_pairwise``?

amueller commented:
The reason why we need to add the intermediate classes into the hierarchy is because we don't add the mixins to the correct (that is left) side, and so we can't have nice things #14044?

Would it be feasible / sensible to actually change the order of mixins and allow overwriting tags? That would simplify a bunch of this, right?

amueller commented:
posted #14635 which will probably simplify this

amueller commented:
ok this has a chance to be merged soon and would make this much easier: #14884

jnothman commented:
With #14884 merged, and a bunch of nitpicks above, are you going to bring this to conclusion, @rok? Thanks.

rok commented:
@jnothman - will try to finish this tonight. Sorry for the delay :)

rok committed:
Switching to overwriting inherited tags.

rok commented:
I'm not sure if I got the general approach right, but I've:
* moved mixins to the left
* replaced intermediate classes by adding `_more_tags` to classes that inherit multioutput flag but should not have it. E.g. `Lars` is multioutput class but `LarsCV` that inherits tags from it is not, so we [adjust it's tags](https://github.com/scikit-learn/scikit-learn/blob/1969e70573eb69fb36a1ef12ef0abdc8636b254e/sklearn/linear_model/least_angle.py#L1361).

Please review :).

Thanks for pushing this @glemaitre @jnothman @amueller!

glemaitre commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1495,6 +1498,71 @@ def check_estimators_partial_fit_n_features(name, estimator_orig):
+                    y_pred[:, i]
+                )
+        elif not tags['poor_score']:
+            y_prob.shape == (n_samples, n_labels)
```
missing assert and error message

glemaitre commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1495,6 +1498,71 @@ def check_estimators_partial_fit_n_features(name, estimator_orig):
 
+@ignore_warnings(category=(DeprecationWarning, FutureWarning))
+def check_classifier_multioutput(name, estimator):
+    n_samples = 42
```
```suggestion
    n_samples, n_labels = 42, 5
```

glemaitre commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1495,6 +1498,71 @@ def check_estimators_partial_fit_n_features(name, estimator_orig):
+@ignore_warnings(category=(DeprecationWarning, FutureWarning))
+def check_classifier_multioutput(name, estimator):
+    n_samples = 42
+    n_labels = 5
```
```suggestion
```

glemaitre commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1495,6 +1498,71 @@ def check_estimators_partial_fit_n_features(name, estimator_orig):
+    estimator.fit(X, y)
+    y_pred = estimator.predict(X)
+
+    assert y_pred.dtype == np.dtype('float')
```
Let's add an error message

```
"Multi-output predictions by a regressor are expected to be floating-point precision. Got {} instead".format(y_pred.dtype)
```

glemaitre replied:
We can also have a similar style as before

```python
assert y_pred.dtype.kind == 'f'
```

rok replied:
Had to set this to `'float64'`.

glemaitre commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1495,6 +1498,71 @@ def check_estimators_partial_fit_n_features(name, estimator_orig):
+    y_pred = estimator.predict(X)
+
+    assert y_pred.dtype == np.dtype('float')
+    assert y_pred.shape == y.shape
```
add an error message as well

glemaitre reviewed (commented):
There is a couple of missing assert in the tests and I would add meaningful error message.
It would ease debugging when rolling your own estimator and that check_estimator is failing.

glemaitre reviewed (changes requested):
So a couple of changes required.

rok commented:
Thanks @glemaitre. I'll do this tonight.

rok committed:
Update sklearn/utils/estimator_checks.py

Co-Authored-By: Guillaume Lemaitre <g.lemaitre58@gmail.com>

rok committed:
Review feedback.

rok commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1495,6 +1498,71 @@ def check_estimators_partial_fit_n_features(name, estimator_orig):
+    estimator.fit(X, y)
+    y_pred = estimator.predict(X)
+
+    assert y_pred.shape == (n_samples, n_labels), y_pred.shape
```
This turns out to be problematic as `assert (a == b, "some msg")` will always be true.

rok commented:
@glemaitre ping :)

glemaitre committed:
Update v0.22.rst

glemaitre commented:
Thanks @rok

rok commented:
Thanks all! :)
